README.md
# music_slider
[![pub package](https://img.shields.io/pub/v/music_slider.svg)](https://pub.dartlang.org/packages/music_slider)

MusicSlider - To Create Music Styled Wavy SeekBar

## preview
<img src="https://raw.githubusercontent.com/RutvikRana/music_slider/main/video_example/giffy.gif" alt="Example App" width="300" height="300">

## What It Can Do?
* Create Default Music Slider
* Create Custom Music Slider By Given Wave Function
* Create Animated Music Slider
* Create ProgressBar By Making detectGesture false
* Create Square Slider Bar
* Create Anything with Wave Function ( I honestly dont know Its capability :) )

## Installation
Follow Installation guide of Pub.dev

## Syntax

1. MusicSlider Class
      ```
      MusicSlider({
      @required this.controller,                                 // Give MusicSliderController To Control MusicSlider
      this.height,                                               // Height
      this.width,                                                // Width
      this.division=30,                                          // Total Number Of Music Bars(Division). Very High Count May Decrease Performance
      this.boxSpace=0.3,                                         // Space Between To Bars ( 0 = No Space, 1 = Full Space)
      this.wave,                                                 // Wave Function To Draw Music Bars
      this.animateWaveByTime = true,                             // Want To Animate It?
      this.tickCount = const Duration(milliseconds: 200),        // Animation Duration between Two Frames
      this.fullBoxFill = true,                                   // Want To Fill Entire Box? If false then Box Fill According To Precise Value
      this.detectGesture=true,                                   // Let user Touch and slide It or not ?
      this.fillColors=const [Colors.red],                        // Fill or Active Bar Colors
      this.emptyColors=const [Colors.blue],                      // Empty or Inactive Bar Colors
      this.initialTime = 0.0                                     // Initial Time For Animation when Starts
      });
      ```
      
#### Important Points:
      * You have to give Controller
      * Give boxSpace 1.0 if you dont want to space in between bars
      * fullBoxFill = false allows to filling by precise value (eg. Not Like Default Filling Entire Box With Color)
      * Make detectGesture = false if you are making ProgressBar
      * You can give any numbers of colors to active/ inactive portion. It will Cycles the List.
      * Dont Forget to make animateWaveByTime = false if you Dont want animated Slider (eg No Time Included In Wavw Function)

### Wave Function
    wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3)
    
    Here, x = Bar Count as X-axis
          t = Time
          a = Amplitude of Wave
     
     - If You Dont Know What Wave Function Is, then Copy Your Desired Function From Example Code.
     
2. MusicController Class
``` 
   MusicSliderController({
      double initialValue = 0.0   // To Give initial Value To Slider. Should Be In Range (0 - 1)
    })
```
### APIs
* double get value                   <br>- Get current value of MusicSlider<br>
``` value = _controller.value```

* seekTo(double value)               <br>- Change Value of MusicSlider<br>
``` _controller.seekTo(0.3)```

 * animateByTime(bool value)         <br>- Start/Stop Animation<br>
 ``` _controller.animateByTime(false)```

 * onSliderChange()                  <br>- Add Callback on SliderChange Event<br>
 ``` _controller.onSliderChange( ()=>print(_controller.value) )```

 * onSliderStart()                   <br>- Add Callback on SliderChange Start Event<br>
 ``` _controller.onSliderStart( ()=>print(_controller.value) )```

 * onSliderEnd()                     <br>- Add Callback on SliderChange End Event<br>
 ``` _controller.onSliderEnd( ()=>print(_controller.value) )```

## Example

```
class _HomeState extends State<Home> {
  MusicSliderController _sliderController = MusicSliderController();

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(title: Text("MusicSlider Example"),),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(padding: EdgeInsets.all(10),
            
              child: MusicSlider(
                emptyColors: [Colors.blue],
                fillColors: [Colors.red],
                controller: _sliderController,
                animateWaveByTime: false,
                height: 50,
                division: 53,
                wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.white],fillColors: [Colors.purple,Colors.blue.shade200,Colors.blue,Colors.green,Colors.yellow,Colors.orange,Colors.red],controller: MusicSliderController(initialValue: 0.5),animateWaveByTime: false,height: 50,division: 53,wave: (x,t,a) => a*cos(x*0.5)*sin(x*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.grey],fillColors: [Colors.purple],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: true,height: 50,division: 53,wave: (x,t,a) => a*sin(x*0.3-t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.blue],fillColors: [Colors.green],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: true,height: 50,division: 53,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.green],fillColors: [Colors.cyan],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: false,height: 50,division: 11,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.blue],fillColors: [Colors.red],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: false,height: 50,division: 11,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),fullBoxFill: false,),),
            RaisedButton(onPressed: (){setState(() { });})
          ]))
      );
  }
}
```

## Contact Me

I Am Rutvik Rana, Medical Student cum Passionate Coder, Invite You To My [Coding(noob to pro)](https://t.me/coding_noob_to_pro) Channel.


CHANGELOG.md
## [0.0.1]

* Landed on Pub.dev

test/music_slider_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:music_slider/musicSlider.dart';

main()=>runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "MusicSlider Test",
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Music Slider Test"),),
      body: SafeArea(child:
        Container(child:
          MusicSlider()
    );
  }
}


lib/musicSlider.dart
///Importing Packages
import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';

///Used To Control MusicSlider
class MusicSliderController{
  double _value;
  Function(double value) _onSliderChange,_onSliderStart,_onSliderEnd;
  Function(bool value) _animateByTime;
  Function _attachedSlider;

  ///Value of slider ( in range of 0 - 1 )
  double get value => _value;

  ///Change the Value of slider ( in range of 0 - 1 )
  void seekTo(double val){
    val = val.clamp(0.0, 1.0);
    if(_attachedSlider == null){
      throw Exception("MusicSliderController Not Attached To any MusicSlider");
    }
    _value = val;
    _attachedSlider();
    if(_onSliderChange != null){_onSliderChange(_value);}
  }

  ///Control Animation Start/Stop
  void animateByTime(bool val){
    if(_animateByTime == null){return;}
    _animateByTime(val);
  }

  ///Callback Called when MusicSlider Change value
  void onSliderChange(Function func){
    _onSliderChange = func;
  }

  ///Callback Called when MusicSlider Change Start
  void onSliderStart(Function func){
    _onSliderStart = func;
  }

  ///Callback Called when MusicSlider Change End
  void onSliderEnd(Function func){
    _onSliderEnd = func;
  }

  ///Constructor
  MusicSliderController({double initialValue = 0.0}){
    _value = initialValue;
  }
}

///MusicSlider Class To Create Wavy SeekBar
class MusicSlider extends StatefulWidget {

  ///Height, Width, Space Between Boxes, Initial Time
  final double height,width,boxSpace,initialTime;
  ///Division
  final int division;
  ///Wave Function
  final double Function(int x,double t,double a) wave;
  ///tickCount ( like Count of frame )
  final Duration tickCount;
  ///animateWaveByTime, fullBoxFill, detectGesture
  final bool animateWaveByTime,fullBoxFill,detectGesture;
  ///Controller
  final MusicSliderController controller;
  ///Colors
  final List<Color> fillColors,emptyColors;

  ///Constructor
  MusicSlider({@required this.controller,this.height,this.width,this.division=10,this.boxSpace=0.3,this.wave,this.animateWaveByTime = true,this.tickCount = const Duration(milliseconds: 200),this.fullBoxFill = true,this.detectGesture=true,this.fillColors=const [Colors.red],this.emptyColors=const [Colors.blue],this.initialTime = 0.0});

  @override
  _MusicSliderState createState() => _MusicSliderState();
}

class _MusicSliderState extends State<MusicSlider> {
  double _time = 0,_width;
  Timer _timer;

  ///initState
  @override
  void initState(){
    super.initState();
    initialize();
  }

  ///initializer
  void initialize(){    
    _time = widget.initialTime;

    if(widget.controller._attachedSlider != null){
      throw Exception("One MusicSliderController Can Attach To Only One MusicSlider");
    }

    widget.controller._attachedSlider = (){setState(() { });};
    widget.controller._animateByTime = (val){
      if(_timer!=null){_timer.cancel();}
      if(val){
        _timer = Timer.periodic(widget.tickCount, (timer){
          setState(() {
            _time++;
          });
        });
      }
    };

    if(_timer!=null){_timer.cancel();}
    if(widget.animateWaveByTime){
      _timer = Timer.periodic(widget.tickCount, (timer){
        setState(() {
          _time++;
        });
      });
    }
  }

  ///Called when widget update by parent
  void didUpdateWidget(oldWidget){
    initialize();
    super.didUpdateWidget(oldWidget);
  }

  ///Click On Slider
  void _sliderTap(TapUpDetails details){
    if(!widget.detectGesture){return;}
    widget.controller.seekTo(details.localPosition.dx/_width);
  }

  ///Drag On Slider Start
  void _hDragStart(DragStartDetails details){
    if(!widget.detectGesture){return;}
    if(widget.controller._onSliderStart == null){return;}
    widget.controller._onSliderStart(widget.controller.value);
  }

  ///Drag On Slider Update
  void _hDragUpdate(DragUpdateDetails details){
    if(!widget.detectGesture){return;}
    widget.controller.seekTo(((details.localPosition.dx)/_width));
  }

  ///Drag On Slider End
  void _hDragEnd(DragEndDetails details){
    if(!widget.detectGesture){return;}
    if(widget.controller._onSliderEnd == null){return;}
    widget.controller._onSliderEnd(widget.controller.value);
  }

  ///Main Build Method
  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
          builder:(c,size){
            double w = widget.width!=null?widget.width:size.biggest.width;
            _width = w;
            double h = widget.height!=null?widget.height:size.biggest.height==null?50:size.biggest.height;
            double _bww = w/widget.division;
            double _bs = (_bww*widget.boxSpace)/2;
            int _passIndex = (widget.controller.value*widget.division).floor();
            double _bw = _bww - _bs*2;
            int l1 = widget.fillColors.length;
            int l2 = widget.emptyColors.length;
            return GestureDetector(
              onTapUp: _sliderTap,
              onHorizontalDragStart: _hDragStart,
              onHorizontalDragUpdate: _hDragUpdate,
              onHorizontalDragEnd: _hDragEnd,
              child: Container(
                color: Colors.transparent,
                height: h,
                width: w,
                child: Row(
                  children: List.generate(widget.division, (index){
                    return Container(
                      margin: EdgeInsets.only(left: _bs,right: _bs),
                      width: _bw,
                      height: widget.wave==null?sin(index).abs()*h:widget.wave(index,_time,h).abs(),
                      color: index<_passIndex?widget.fillColors[index%l1]:widget.emptyColors[index%l2],
                      child: index==_passIndex && !widget.fullBoxFill?Align(alignment: Alignment.centerLeft,child: Container(width: ((w*widget.controller.value)-(index*_bww)),color: Colors.red,),):null,);
                  }),
                ),
      ),
            );},
    );
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'musicSlider.dart';

void main() => runApp(MyApp());

// ------ Root Widget ---------
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "MusicSlider Example",
      theme: ThemeData(
          primarySwatch: Colors.green,
          canvasColor: Colors.blue.shade100,
          visualDensity: VisualDensity.adaptivePlatformDensity,
          platform: TargetPlatform.android),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  MusicSliderController _sliderController = MusicSliderController();

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(title: Text("MusicSlider Example"),),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(padding: EdgeInsets.all(10),
            
              child: MusicSlider(
                emptyColors: [Colors.blue],
                fillColors: [Colors.red],
                controller: _sliderController,
                animateWaveByTime: false,
                height: 50,
                division: 53,
                wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.white],fillColors: [Colors.purple,Colors.blue.shade200,Colors.blue,Colors.green,Colors.yellow,Colors.orange,Colors.red],controller: MusicSliderController(initialValue: 0.5),animateWaveByTime: false,height: 50,division: 53,wave: (x,t,a) => a*cos(x*0.5)*sin(x*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.grey],fillColors: [Colors.purple],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: true,height: 50,division: 53,wave: (x,t,a) => a*sin(x*0.3-t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.blue],fillColors: [Colors.green],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: true,height: 50,division: 53,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.green],fillColors: [Colors.cyan],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: false,height: 50,division: 11,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),),),
            Padding(padding: EdgeInsets.all(10),child: MusicSlider(emptyColors: [Colors.blue],fillColors: [Colors.red],controller: MusicSliderController(initialValue: 0.2),animateWaveByTime: false,height: 50,division: 11,wave: (x,t,a) => a*sin(x*0.3-t*0.3)*sin(x*0.3+t*0.3),fullBoxFill: false,),),
            RaisedButton(onPressed: (){setState(() { });})
          ]))
      );
  }
}
